{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "emihu_gff = './Emihu1_all_genes.gff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = ['scaffold', 'jgi', 'type', 'start', 'end', 'direction1', 'direction2', 'phase', 'infos']\n",
    "\n",
    "# Load the GFF file\n",
    "df_gff = pd.read_csv(emihu_gff, sep='\\t', comment='#', names=columns)\n",
    "\n",
    "# Split the 'attributes' column by ';' into multiple columns\n",
    "df_gff_split = df_gff['infos'].str.split(';', expand=True)\n",
    "# Rename the new columns (optional)\n",
    "df_gff_split.columns = ['name', 'proteinId', 'n_exon']\n",
    "df_gff.columns = df_gff.columns.str.strip()\n",
    "\n",
    "# Add the split columns back to the original DataFrame\n",
    "df_gff = pd.concat([df_gff, df_gff_split], axis=1)\n",
    "df_gff.drop(['infos'], axis=1, inplace=True)\n",
    "\n",
    "df_gff = df_gff[~df_gff['proteinId'].isna()]\n",
    "# convert CDS position to integer and subtract by 1 because string start from 0 in python\n",
    "df_gff['start'] = df_gff['start'].astype(int) - 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gff = df_gff[~df_gff['proteinId'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>jgi</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>direction1</th>\n",
       "      <th>direction2</th>\n",
       "      <th>phase</th>\n",
       "      <th>name</th>\n",
       "      <th>proteinId</th>\n",
       "      <th>n_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>exon</td>\n",
       "      <td>203</td>\n",
       "      <td>287</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name \"gm1.100001\"</td>\n",
       "      <td>transcriptId 193701</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>203</td>\n",
       "      <td>287</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gm1.100001\"</td>\n",
       "      <td>proteinId 193701</td>\n",
       "      <td>exonNumber 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>exon</td>\n",
       "      <td>1701</td>\n",
       "      <td>1737</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name \"gm1.100001\"</td>\n",
       "      <td>transcriptId 193701</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1701</td>\n",
       "      <td>1737</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gm1.100001\"</td>\n",
       "      <td>proteinId 193701</td>\n",
       "      <td>exonNumber 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>exon</td>\n",
       "      <td>1939</td>\n",
       "      <td>2059</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name \"gm1.100001\"</td>\n",
       "      <td>transcriptId 193701</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2076640</th>\n",
       "      <td>scaffold_999</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>12061</td>\n",
       "      <td>12208</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"fgenesh_newKGs_pm.999__2\"</td>\n",
       "      <td>proteinId 317413</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2076642</th>\n",
       "      <td>scaffold_999</td>\n",
       "      <td>JGI</td>\n",
       "      <td>exon</td>\n",
       "      <td>12294</td>\n",
       "      <td>12343</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name \"fgenesh_newKGs_pm.999__2\"</td>\n",
       "      <td>transcriptId 317413</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2076643</th>\n",
       "      <td>scaffold_999</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>12294</td>\n",
       "      <td>12343</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"fgenesh_newKGs_pm.999__2\"</td>\n",
       "      <td>proteinId 317413</td>\n",
       "      <td>exonNumber 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2076644</th>\n",
       "      <td>scaffold_999</td>\n",
       "      <td>JGI</td>\n",
       "      <td>exon</td>\n",
       "      <td>12548</td>\n",
       "      <td>12760</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name \"fgenesh_newKGs_pm.999__2\"</td>\n",
       "      <td>transcriptId 317413</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2076645</th>\n",
       "      <td>scaffold_999</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>12548</td>\n",
       "      <td>12760</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>name \"fgenesh_newKGs_pm.999__2\"</td>\n",
       "      <td>proteinId 317413</td>\n",
       "      <td>exonNumber 3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1690186 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             scaffold  jgi  type  start    end direction1 direction2 phase  \\\n",
       "0          scaffold_1  JGI  exon    203    287          .          -     .   \n",
       "1          scaffold_1  JGI   CDS    203    287          .          -     0   \n",
       "3          scaffold_1  JGI  exon   1701   1737          .          -     .   \n",
       "4          scaffold_1  JGI   CDS   1701   1737          .          -     0   \n",
       "5          scaffold_1  JGI  exon   1939   2059          .          -     .   \n",
       "...               ...  ...   ...    ...    ...        ...        ...   ...   \n",
       "2076640  scaffold_999  JGI   CDS  12061  12208          .          +     0   \n",
       "2076642  scaffold_999  JGI  exon  12294  12343          .          +     .   \n",
       "2076643  scaffold_999  JGI   CDS  12294  12343          .          +     0   \n",
       "2076644  scaffold_999  JGI  exon  12548  12760          .          +     .   \n",
       "2076645  scaffold_999  JGI   CDS  12548  12760          .          +     1   \n",
       "\n",
       "                                    name             proteinId         n_exon  \n",
       "0                      name \"gm1.100001\"   transcriptId 193701           None  \n",
       "1                      name \"gm1.100001\"      proteinId 193701   exonNumber 3  \n",
       "3                      name \"gm1.100001\"   transcriptId 193701           None  \n",
       "4                      name \"gm1.100001\"      proteinId 193701   exonNumber 2  \n",
       "5                      name \"gm1.100001\"   transcriptId 193701           None  \n",
       "...                                  ...                   ...            ...  \n",
       "2076640  name \"fgenesh_newKGs_pm.999__2\"      proteinId 317413   exonNumber 1  \n",
       "2076642  name \"fgenesh_newKGs_pm.999__2\"   transcriptId 317413           None  \n",
       "2076643  name \"fgenesh_newKGs_pm.999__2\"      proteinId 317413   exonNumber 2  \n",
       "2076644  name \"fgenesh_newKGs_pm.999__2\"   transcriptId 317413           None  \n",
       "2076645  name \"fgenesh_newKGs_pm.999__2\"      proteinId 317413   exonNumber 3  \n",
       "\n",
       "[1690186 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## check integrity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "scaffold\n",
       "scaffold_1       32226\n",
       "scaffold_5       28201\n",
       "scaffold_3       26272\n",
       "scaffold_4       23537\n",
       "scaffold_7       16370\n",
       "                 ...  \n",
       "scaffold_6462        2\n",
       "scaffold_4460        2\n",
       "scaffold_7819        2\n",
       "scaffold_9088        2\n",
       "scaffold_9683        2\n",
       "Name: count, Length: 7760, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gff['scaffold'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>jgi</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>direction1</th>\n",
       "      <th>direction2</th>\n",
       "      <th>phase</th>\n",
       "      <th>name</th>\n",
       "      <th>proteinId</th>\n",
       "      <th>n_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>413483</th>\n",
       "      <td>scaffold_5</td>\n",
       "      <td>JGI</td>\n",
       "      <td>exon</td>\n",
       "      <td>243</td>\n",
       "      <td>654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name \"gm1.500001\"</td>\n",
       "      <td>transcriptId 196822</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413484</th>\n",
       "      <td>scaffold_5</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>243</td>\n",
       "      <td>654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gm1.500001\"</td>\n",
       "      <td>proteinId 196822</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413487</th>\n",
       "      <td>scaffold_5</td>\n",
       "      <td>JGI</td>\n",
       "      <td>exon</td>\n",
       "      <td>739</td>\n",
       "      <td>763</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name \"gm1.500002\"</td>\n",
       "      <td>transcriptId 196823</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413488</th>\n",
       "      <td>scaffold_5</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>739</td>\n",
       "      <td>763</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gm1.500002\"</td>\n",
       "      <td>proteinId 196823</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>413490</th>\n",
       "      <td>scaffold_5</td>\n",
       "      <td>JGI</td>\n",
       "      <td>exon</td>\n",
       "      <td>1105</td>\n",
       "      <td>1204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>name \"gm1.500002\"</td>\n",
       "      <td>transcriptId 196823</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2057058</th>\n",
       "      <td>scaffold_5</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2128589</td>\n",
       "      <td>2128617</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"fgenesh_newKGs_pm.5__155\"</td>\n",
       "      <td>proteinId 308753</td>\n",
       "      <td>exonNumber 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2057059</th>\n",
       "      <td>scaffold_5</td>\n",
       "      <td>JGI</td>\n",
       "      <td>exon</td>\n",
       "      <td>2128705</td>\n",
       "      <td>2128815</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name \"fgenesh_newKGs_pm.5__155\"</td>\n",
       "      <td>transcriptId 308753</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2057060</th>\n",
       "      <td>scaffold_5</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2128705</td>\n",
       "      <td>2128815</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>name \"fgenesh_newKGs_pm.5__155\"</td>\n",
       "      <td>proteinId 308753</td>\n",
       "      <td>exonNumber 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2057061</th>\n",
       "      <td>scaffold_5</td>\n",
       "      <td>JGI</td>\n",
       "      <td>exon</td>\n",
       "      <td>2128887</td>\n",
       "      <td>2128971</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>name \"fgenesh_newKGs_pm.5__155\"</td>\n",
       "      <td>transcriptId 308753</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2057062</th>\n",
       "      <td>scaffold_5</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2128887</td>\n",
       "      <td>2128971</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"fgenesh_newKGs_pm.5__155\"</td>\n",
       "      <td>proteinId 308753</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28201 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           scaffold  jgi  type    start      end direction1 direction2 phase  \\\n",
       "413483   scaffold_5  JGI  exon      243      654          .          -     .   \n",
       "413484   scaffold_5  JGI   CDS      243      654          .          -     0   \n",
       "413487   scaffold_5  JGI  exon      739      763          .          +     .   \n",
       "413488   scaffold_5  JGI   CDS      739      763          .          +     0   \n",
       "413490   scaffold_5  JGI  exon     1105     1204          .          +     .   \n",
       "...             ...  ...   ...      ...      ...        ...        ...   ...   \n",
       "2057058  scaffold_5  JGI   CDS  2128589  2128617          .          -     0   \n",
       "2057059  scaffold_5  JGI  exon  2128705  2128815          .          -     .   \n",
       "2057060  scaffold_5  JGI   CDS  2128705  2128815          .          -     1   \n",
       "2057061  scaffold_5  JGI  exon  2128887  2128971          .          -     .   \n",
       "2057062  scaffold_5  JGI   CDS  2128887  2128971          .          -     0   \n",
       "\n",
       "                                    name             proteinId         n_exon  \n",
       "413483                 name \"gm1.500001\"   transcriptId 196822           None  \n",
       "413484                 name \"gm1.500001\"      proteinId 196822   exonNumber 1  \n",
       "413487                 name \"gm1.500002\"   transcriptId 196823           None  \n",
       "413488                 name \"gm1.500002\"      proteinId 196823   exonNumber 1  \n",
       "413490                 name \"gm1.500002\"   transcriptId 196823           None  \n",
       "...                                  ...                   ...            ...  \n",
       "2057058  name \"fgenesh_newKGs_pm.5__155\"      proteinId 308753   exonNumber 3  \n",
       "2057059  name \"fgenesh_newKGs_pm.5__155\"   transcriptId 308753           None  \n",
       "2057060  name \"fgenesh_newKGs_pm.5__155\"      proteinId 308753   exonNumber 2  \n",
       "2057061  name \"fgenesh_newKGs_pm.5__155\"   transcriptId 308753           None  \n",
       "2057062  name \"fgenesh_newKGs_pm.5__155\"      proteinId 308753   exonNumber 1  \n",
       "\n",
       "[28201 rows x 11 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gff[df_gff['scaffold'] == 'scaffold_5']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2v/0h890r5d3g93n2j6tb6k9cf00000gn/T/ipykernel_87031/2718034935.py:2: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  result_53 = df_gff.groupby('name').apply(\n",
      "/var/folders/2v/0h890r5d3g93n2j6tb6k9cf00000gn/T/ipykernel_87031/2718034935.py:9: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  result_35 = df_gff.groupby('name').apply(\n"
     ]
    }
   ],
   "source": [
    "# get the first cds of protein which is translated by 5-3\n",
    "result_53 = df_gff.groupby('name').apply(\n",
    "    lambda group: group.loc[\n",
    "        (group['type'] == 'CDS') & (group['direction2'] == '+')\n",
    "    ].head(1)  # get first match\n",
    ").reset_index(drop=True)\n",
    "\n",
    "# get the last cds of protein which is translated by 3-5\n",
    "result_35 = df_gff.groupby('name').apply(\n",
    "    lambda group: group.loc[\n",
    "        (group['type'] == 'CDS') & (group['direction2'] == '-')\n",
    "    ].iloc[-1] if any((group['type'] == 'CDS') & (group['direction2'] == '-')) else None\n",
    ").dropna().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>jgi</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>direction1</th>\n",
       "      <th>direction2</th>\n",
       "      <th>phase</th>\n",
       "      <th>name</th>\n",
       "      <th>proteinId</th>\n",
       "      <th>n_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1273784.0</td>\n",
       "      <td>1273873.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.100.1\"</td>\n",
       "      <td>proteinId 60925</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1238222.0</td>\n",
       "      <td>1238786.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.102.1\"</td>\n",
       "      <td>proteinId 61047</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1999638.0</td>\n",
       "      <td>2000029.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.103.1\"</td>\n",
       "      <td>proteinId 61109</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>857135.0</td>\n",
       "      <td>857702.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.105.1\"</td>\n",
       "      <td>proteinId 60920</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1189068.0</td>\n",
       "      <td>1189311.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.106.1\"</td>\n",
       "      <td>proteinId 60980</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113277</th>\n",
       "      <td>scaffold_998</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>5716.0</td>\n",
       "      <td>5846.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>name \"gw1.998.2.1\"</td>\n",
       "      <td>proteinId 46252</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113278</th>\n",
       "      <td>scaffold_998</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>5716.0</td>\n",
       "      <td>5858.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>name \"gw1.998.3.1\"</td>\n",
       "      <td>proteinId 46444</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113279</th>\n",
       "      <td>scaffold_998</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>67.0</td>\n",
       "      <td>379.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gw1.998.4.1\"</td>\n",
       "      <td>proteinId 59951</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113280</th>\n",
       "      <td>scaffold_998</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>67.0</td>\n",
       "      <td>379.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gw1.998.5.1\"</td>\n",
       "      <td>proteinId 59974</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113281</th>\n",
       "      <td>scaffold_998</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>58.0</td>\n",
       "      <td>331.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gw1.998.8.1\"</td>\n",
       "      <td>proteinId 60282</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>228488 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            scaffold  jgi type      start        end direction1 direction2  \\\n",
       "0         scaffold_1  JGI  CDS  1273784.0  1273873.0          .          +   \n",
       "1         scaffold_1  JGI  CDS  1238222.0  1238786.0          .          +   \n",
       "2         scaffold_1  JGI  CDS  1999638.0  2000029.0          .          +   \n",
       "3         scaffold_1  JGI  CDS   857135.0   857702.0          .          +   \n",
       "4         scaffold_1  JGI  CDS  1189068.0  1189311.0          .          +   \n",
       "...              ...  ...  ...        ...        ...        ...        ...   \n",
       "113277  scaffold_998  JGI  CDS     5716.0     5846.0          .          -   \n",
       "113278  scaffold_998  JGI  CDS     5716.0     5858.0          .          -   \n",
       "113279  scaffold_998  JGI  CDS       67.0      379.0          .          -   \n",
       "113280  scaffold_998  JGI  CDS       67.0      379.0          .          -   \n",
       "113281  scaffold_998  JGI  CDS       58.0      331.0          .          -   \n",
       "\n",
       "       phase                  name         proteinId         n_exon  \n",
       "0          0  name \"e_gw1.1.100.1\"   proteinId 60925   exonNumber 1  \n",
       "1          0  name \"e_gw1.1.102.1\"   proteinId 61047   exonNumber 1  \n",
       "2          0  name \"e_gw1.1.103.1\"   proteinId 61109   exonNumber 1  \n",
       "3          0  name \"e_gw1.1.105.1\"   proteinId 60920   exonNumber 1  \n",
       "4          0  name \"e_gw1.1.106.1\"   proteinId 60980   exonNumber 1  \n",
       "...      ...                   ...               ...            ...  \n",
       "113277     2    name \"gw1.998.2.1\"   proteinId 46252   exonNumber 1  \n",
       "113278     2    name \"gw1.998.3.1\"   proteinId 46444   exonNumber 1  \n",
       "113279     0    name \"gw1.998.4.1\"   proteinId 59951   exonNumber 1  \n",
       "113280     0    name \"gw1.998.5.1\"   proteinId 59974   exonNumber 1  \n",
       "113281     0    name \"gw1.998.8.1\"   proteinId 60282   exonNumber 1  \n",
       "\n",
       "[228488 rows x 11 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = pd.concat((result_53, result_35))\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "scaffold\n",
       "scaffold_1       2096\n",
       "scaffold_5       1642\n",
       "scaffold_4       1591\n",
       "scaffold_3       1579\n",
       "scaffold_7       1031\n",
       "                 ... \n",
       "scaffold_3133       1\n",
       "scaffold_3129       1\n",
       "scaffold_3124       1\n",
       "scaffold_3123       1\n",
       "scaffold_9996       1\n",
       "Name: count, Length: 5247, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.loc[result['direction2'] == '-']['scaffold'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fasta Sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "emihu_seq_path = './Emihu1_masked_scaffolds.fasta'\n",
    "emihu_seq_dict = {}\n",
    "key = ''\n",
    "with open(emihu_seq_path, 'r') as f:\n",
    "    for line in f.readlines():\n",
    "        if \">\" in line:\n",
    "            # save scaffold number as dict key\n",
    "            key = line[1:].strip()  # avoid '>'\n",
    "            # create a value for storing the sequence\n",
    "            emihu_seq_dict[key] = ''\n",
    "        else:\n",
    "            emihu_seq_dict[key] += line.strip()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## verify"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>jgi</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>direction1</th>\n",
       "      <th>direction2</th>\n",
       "      <th>phase</th>\n",
       "      <th>name</th>\n",
       "      <th>proteinId</th>\n",
       "      <th>n_exon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>107078</th>\n",
       "      <td>scaffold_13</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>227696.0</td>\n",
       "      <td>227915.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gw1.13.147.1\"</td>\n",
       "      <td>proteinId 59809</td>\n",
       "      <td>exonNumber 1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           scaffold  jgi type     start       end direction1 direction2 phase  \\\n",
       "107078  scaffold_13  JGI  CDS  227696.0  227915.0          .          +     0   \n",
       "\n",
       "                       name         proteinId         n_exon  \n",
       "107078  name \"gw1.13.147.1\"   proteinId 59809   exonNumber 1  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result[result['name'] == 'name \"gw1.13.147.1\"']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2v/0h890r5d3g93n2j6tb6k9cf00000gn/T/ipykernel_87031/2129101991.py:2: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  int(result[result['name'] == 'name \"gw1.13.147.1\"']['start']):int(result[result['name'] == 'name \"gw1.13.147.1\"']['end'])\n"
     ]
    }
   ],
   "source": [
    "tmp = emihu_seq_dict['scaffold_13'][\n",
    "    int(result[result['name'] == 'name \"gw1.13.147.1\"']['start']):int(result[result['name'] == 'name \"gw1.13.147.1\"']['end'])\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(219,\n",
       " 'GGTCCCGTCATCGGCATCGACCTCGGCACCACCTATTCCTGCGTCGGAGTCTACCGGAAGGGCAAGGTGGAGATCATCGCCAACGAGCAGGGGAACCGAGTGACGCCGTCGTGGGTGGCGTTCACCGACGACGGCGAGCGGCTGGTCGGCGACGCCGCGCGCTCGCAGTCCTCGCTCAACCCGGTAAACACGATTTACGACGCGAAGCGGCTCATCGGC')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(tmp), tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='red'> Same result shown on JGI webpage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find the closest M to * in the inverse direction from CDS to *   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAIVMGR*KGAR*\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Example DNA sequence\n",
    "dna_seq = Seq(\"ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\")\n",
    "\n",
    "# Translate to protein\n",
    "protein_seq = dna_seq.translate()\n",
    "\n",
    "print(protein_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reVerifyM(\n",
    "        fasta_seq: str, \n",
    "        cds_start: float,\n",
    "        cds_end:float,\n",
    "        direction:str,\n",
    "        ) -> int:\n",
    "    \"\"\"\n",
    "    Take an input of fasta acide amine sequence, CDS position from JGI \n",
    "    and re-find the M position in the given sequence\n",
    "    \"\"\"\n",
    "    cds_start = int(cds_start)\n",
    "    if cds_start < 0 or cds_start >= len(fasta_seq) or cds_start > cds_end:\n",
    "        return cds_start\n",
    "\n",
    "    cds_end = int(cds_end)\n",
    "    if cds_end < 0 or cds_end >= len(fasta_seq):\n",
    "        return cds_start\n",
    "    # Forward\n",
    "    if direction == \"+\":\n",
    "        # translate DNA to AA\n",
    "        # include the first translated aa\n",
    "        sub_seq = fasta_seq[:cds_start+3]\n",
    "        # identify the translation framework: +0,+1 or +2 depend on the length of sequence\n",
    "        framework = len(sub_seq)%3\n",
    "        sub_seq = str(Seq(sub_seq[framework:]).translate())\n",
    "        # find the closest codon stop\n",
    "        # consider the closest * near the cds start position so inverse make the search run faster\n",
    "        inversed_seq = sub_seq[::-1]\n",
    "        match = re.search(r'\\*', inversed_seq)\n",
    "        # the first codon stop encoutered\n",
    "        stop_pos = match.start() if match else len(sub_seq)\n",
    "\n",
    "        # find the closest aa M to the first * by slicing again the inversed aa sequence\n",
    "        tmp_seq = inversed_seq[:stop_pos]\n",
    "        # find the first M encoutered\n",
    "        M_pos = tmp_seq.find('M')\n",
    "        if M_pos >= 0:\n",
    "            M_pos = (len(inversed_seq) - M_pos - 1)*3 + framework\n",
    "        return M_pos\n",
    "    # Backward\n",
    "    else:\n",
    "        # find the true (furthest) M position by going backward the CDS sequence\n",
    "        sub_seq = fasta_seq[cds_start-3: cds_end]\n",
    "        sub_seq = str(Seq(sub_seq).reverse_complement().translate())\n",
    "        # print(sub_seq)\n",
    "        # find the last M encoutered\n",
    "        inversed_seq = sub_seq[::-1]\n",
    "        M_pos = inversed_seq.find('M')\n",
    "        if M_pos >= 0:\n",
    "            M_pos = len(inversed_seq) - M_pos - 1\n",
    "            # print(sub_seq[M_pos:])\n",
    "            # print(cds_end - M_pos*3)\n",
    "            # print(Seq(fasta_seq[cds_start:cds_end-M_pos*3]).reverse_complement().translate())\n",
    "            return cds_end - M_pos*3\n",
    "        return int(M_pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GGTCCCGTCATCGGCATCGACCTCGGCACCACCTATTCCTGCGTCGGAGTCTACCGGAAGGGCAAGGTGGAGATCATCGCCAACGAGCAGGGGAACCGAGTGACGCCGTCGTGGGTGGCGTTCACCGACGACGGCGAGCGGCTGGTCGGCGACGCCGCGCGCTCGCAGTCCTCGCTCAACCCGGTAAACACGATTTACGACGCGAAGCGGCTCATCGGC'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GPVIGIDLGTTYSCVGVYRKGKVEIIANEQGNRVTPSWVAFTDDGERLVGDAAR...LIG')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Seq(tmp).translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2v/0h890r5d3g93n2j6tb6k9cf00000gn/T/ipykernel_87031/2200201372.py:1: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  len(emihu_seq_dict[result[result['name'] == 'name \"gw1.13.147.1\"']['scaffold'].iloc[0]][:int(result[result['name'] == 'name \"gw1.13.147.1\"']['start'])])\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "227696"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(emihu_seq_dict[result[result['name'] == 'name \"gw1.13.147.1\"']['scaffold'].iloc[0]][:int(result[result['name'] == 'name \"gw1.13.147.1\"']['start'])])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(75898, 2)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "227696 // 3, 227696 % 3,  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cds_pos = int(result.loc[result['name'] == 'name \"gw1.13.147.1\"', 'start'].iloc[0])\n",
    "# cds_end = int(result.loc[result['name'] == 'name \"gw1.13.147.1\"', 'end'].iloc[0])\n",
    "# scaffold = result.loc[result['name'] == 'name \"gw1.13.147.1\"', 'scaffold'].iloc[0]\n",
    "# print(Seq(emihu_seq_dict[scaffold][cds_pos:cds_end]).translate())\n",
    "# m_pos = reVerifyM(emihu_seq_dict[scaffold],\n",
    "#       cds_pos=cds_pos\n",
    "# )\n",
    "# m_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seq(emihu_seq_dict[result[result['name'] == 'name \"gw1.13.147.1\"']['scaffold'].iloc[0]][\n",
    "#     m_pos:int(result.loc[result['name'] == 'name \"gw1.13.147.1\"', 'end'].iloc[0])\n",
    "# ]).translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# result = result[~result['proteinId'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# result[result['proteinId'].str.contains('421892')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "166785\n",
      "MRLLITLIPLLAAAATAEEAEAVGPVIGIDLGTTYSCVGVYRKGKVEIIANEQGNRVTPSWVAFTDDGERLVGDAARSQSSLNPVNTIYDAKRLIGRGFRDEEVQKDAEHWPFKVVATAEGKPAVEVAAGGATKRLLPQEISAMVLGKMKG\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Seq('MRLLITLIPLLAAAATAEEAEAVGPVIGIDLGTTYSCVGVYRKGKVEIIANEQG...MKG')"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cds_pos = int(result.loc[result['name'] == 'name \"estExtDG_fgenesh_newKGs_pm.C_470006\"', 'start'].iloc[0])\n",
    "# m_pos = reVerifyM(emihu_seq_dict[result.loc[result['name'] == 'name \"estExtDG_fgenesh_newKGs_pm.C_470006\"', 'scaffold'].iloc[0]],\n",
    "#       cds_pos=cds_pos\n",
    "# )\n",
    "# print(m_pos)\n",
    "# print(Seq(emihu_seq_dict[result.loc[result['name'] == 'name \"estExtDG_fgenesh_newKGs_pm.C_470006\"', 'scaffold'].iloc[0]][\n",
    "#     int(result.loc[result['name'] == 'name \"estExtDG_fgenesh_newKGs_pm.C_470006\"', 'start'].iloc[0]):int(result.loc[result['name'] == 'name \"estExtDG_fgenesh_newKGs_pm.C_470006\"', 'end'].iloc[0])\n",
    "# ]).translate())\n",
    "# Seq(emihu_seq_dict[result.loc[result['name'] == 'name \"estExtDG_fgenesh_newKGs_pm.C_470006\"', 'scaffold'].iloc[0]][\n",
    "#     m_pos:int(result.loc[result['name'] == 'name \"estExtDG_fgenesh_newKGs_pm.C_470006\"', 'end'].iloc[0])\n",
    "# ]).translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tranquocan.truong/Documents/AnTTQ/U_Sorbonne/M1S2/Projet/projectS1/lib/python3.9/site-packages/Bio/Seq.py:2879: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "results = []\n",
    "\n",
    "for scaffold, group_df in result.groupby(['scaffold']):\n",
    "    fasta_seq = emihu_seq_dict[scaffold[0]]  # get once\n",
    "    starts = group_df['start'].astype(int).tolist()\n",
    "    ends = group_df['end'].astype(int).tolist()\n",
    "    directions = group_df['direction2'].tolist()\n",
    "\n",
    "    # Vectorized by list comprehension per group\n",
    "    true_M_positions = [reVerifyM(fasta_seq, cds_start, cds_end, direction) for (cds_start, cds_end, direction) in zip(starts, ends, directions)]\n",
    "\n",
    "    group_result = group_df.copy()\n",
    "    group_result['true_M_pos'] = true_M_positions\n",
    "    group_result['no_M'] = False\n",
    "    group_result.loc[group_result['true_M_pos'] == -1, 'no_M'] = True \n",
    "\n",
    "    # calculate the differences\n",
    "    group_result['M_pos_diff'] = 0\n",
    "    # only gene in which has M can calculate the difference\n",
    "    mask = ~group_result['no_M']\n",
    "    \n",
    "    # forward because of direction +\n",
    "    mask2 = group_result['direction2'] == '+'\n",
    "    group_result.loc[mask & mask2, 'M_pos_diff'] = (\n",
    "        group_result.loc[mask & mask2, 'start'] - group_result.loc[mask & mask2, 'true_M_pos']\n",
    "    ).abs()\n",
    "\n",
    "    # backward because of direction -\n",
    "    mask3 = group_result['direction2'] == '-'\n",
    "    group_result.loc[mask & mask3, 'M_pos_diff'] = (\n",
    "        group_result.loc[mask & mask3, 'end'] - group_result.loc[mask & mask3, 'true_M_pos']\n",
    "    ).abs()\n",
    "    results.append(group_result)\n",
    "    break \n",
    "\n",
    "# Concatenate all groups\n",
    "result = pd.concat(results, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>jgi</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>direction1</th>\n",
       "      <th>direction2</th>\n",
       "      <th>phase</th>\n",
       "      <th>name</th>\n",
       "      <th>proteinId</th>\n",
       "      <th>n_exon</th>\n",
       "      <th>true_M_pos</th>\n",
       "      <th>no_M</th>\n",
       "      <th>M_pos_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1273784.0</td>\n",
       "      <td>1273873.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.100.1\"</td>\n",
       "      <td>proteinId 60925</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1238222.0</td>\n",
       "      <td>1238786.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.102.1\"</td>\n",
       "      <td>proteinId 61047</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1999638.0</td>\n",
       "      <td>2000029.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.103.1\"</td>\n",
       "      <td>proteinId 61109</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>1999638</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>857135.0</td>\n",
       "      <td>857702.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.105.1\"</td>\n",
       "      <td>proteinId 60920</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>857135</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1189068.0</td>\n",
       "      <td>1189311.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.106.1\"</td>\n",
       "      <td>proteinId 60980</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4190</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1456659.0</td>\n",
       "      <td>1456868.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>name \"gw1.1.87.1\"</td>\n",
       "      <td>proteinId 45700</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>1456784</td>\n",
       "      <td>False</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4191</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2843908.0</td>\n",
       "      <td>2844120.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>name \"gw1.1.89.1\"</td>\n",
       "      <td>proteinId 45750</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>2844069</td>\n",
       "      <td>False</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4192</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2705691.0</td>\n",
       "      <td>2705999.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>name \"gw1.1.90.1\"</td>\n",
       "      <td>proteinId 45766</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>2705984</td>\n",
       "      <td>False</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4193</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1629617.0</td>\n",
       "      <td>1629799.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>name \"gw1.1.92.1\"</td>\n",
       "      <td>proteinId 46190</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>1629625</td>\n",
       "      <td>False</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4194</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1254079.0</td>\n",
       "      <td>1254451.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gw1.1.98.1\"</td>\n",
       "      <td>proteinId 46409</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>1254193</td>\n",
       "      <td>False</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4195 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        scaffold  jgi type      start        end direction1 direction2 phase  \\\n",
       "0     scaffold_1  JGI  CDS  1273784.0  1273873.0          .          +     0   \n",
       "1     scaffold_1  JGI  CDS  1238222.0  1238786.0          .          +     0   \n",
       "2     scaffold_1  JGI  CDS  1999638.0  2000029.0          .          +     0   \n",
       "3     scaffold_1  JGI  CDS   857135.0   857702.0          .          +     0   \n",
       "4     scaffold_1  JGI  CDS  1189068.0  1189311.0          .          +     0   \n",
       "...          ...  ...  ...        ...        ...        ...        ...   ...   \n",
       "4190  scaffold_1  JGI  CDS  1456659.0  1456868.0          .          -     1   \n",
       "4191  scaffold_1  JGI  CDS  2843908.0  2844120.0          .          -     1   \n",
       "4192  scaffold_1  JGI  CDS  2705691.0  2705999.0          .          -     1   \n",
       "4193  scaffold_1  JGI  CDS  1629617.0  1629799.0          .          -     1   \n",
       "4194  scaffold_1  JGI  CDS  1254079.0  1254451.0          .          -     0   \n",
       "\n",
       "                      name         proteinId         n_exon  true_M_pos  \\\n",
       "0     name \"e_gw1.1.100.1\"   proteinId 60925   exonNumber 1          -1   \n",
       "1     name \"e_gw1.1.102.1\"   proteinId 61047   exonNumber 1          -1   \n",
       "2     name \"e_gw1.1.103.1\"   proteinId 61109   exonNumber 1     1999638   \n",
       "3     name \"e_gw1.1.105.1\"   proteinId 60920   exonNumber 1      857135   \n",
       "4     name \"e_gw1.1.106.1\"   proteinId 60980   exonNumber 1          -1   \n",
       "...                    ...               ...            ...         ...   \n",
       "4190     name \"gw1.1.87.1\"   proteinId 45700   exonNumber 1     1456784   \n",
       "4191     name \"gw1.1.89.1\"   proteinId 45750   exonNumber 1     2844069   \n",
       "4192     name \"gw1.1.90.1\"   proteinId 45766   exonNumber 1     2705984   \n",
       "4193     name \"gw1.1.92.1\"   proteinId 46190   exonNumber 1     1629625   \n",
       "4194     name \"gw1.1.98.1\"   proteinId 46409   exonNumber 1     1254193   \n",
       "\n",
       "       no_M  M_pos_diff  \n",
       "0      True           0  \n",
       "1      True           0  \n",
       "2     False           0  \n",
       "3     False           0  \n",
       "4      True           0  \n",
       "...     ...         ...  \n",
       "4190  False          84  \n",
       "4191  False          51  \n",
       "4192  False          15  \n",
       "4193  False         174  \n",
       "4194  False         258  \n",
       "\n",
       "[4195 rows x 14 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>jgi</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>direction1</th>\n",
       "      <th>direction2</th>\n",
       "      <th>phase</th>\n",
       "      <th>name</th>\n",
       "      <th>proteinId</th>\n",
       "      <th>n_exon</th>\n",
       "      <th>true_M_pos</th>\n",
       "      <th>no_M</th>\n",
       "      <th>M_pos_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1273784.0</td>\n",
       "      <td>1273873.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.100.1\"</td>\n",
       "      <td>proteinId 60925</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1238222.0</td>\n",
       "      <td>1238786.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.102.1\"</td>\n",
       "      <td>proteinId 61047</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1999638.0</td>\n",
       "      <td>2000029.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.103.1\"</td>\n",
       "      <td>proteinId 61109</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>1999638</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>857135.0</td>\n",
       "      <td>857702.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.105.1\"</td>\n",
       "      <td>proteinId 60920</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>857135</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1189068.0</td>\n",
       "      <td>1189311.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"e_gw1.1.106.1\"</td>\n",
       "      <td>proteinId 60980</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2094</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1394773.0</td>\n",
       "      <td>1395190.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gw1.1.94.1\"</td>\n",
       "      <td>proteinId 46209</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>1394773</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2095</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2885127.0</td>\n",
       "      <td>2885541.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gw1.1.95.1\"</td>\n",
       "      <td>proteinId 46224</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2096</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1898490.0</td>\n",
       "      <td>1898835.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gw1.1.96.1\"</td>\n",
       "      <td>proteinId 46246</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>1898307</td>\n",
       "      <td>False</td>\n",
       "      <td>183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2097</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>718868.0</td>\n",
       "      <td>719048.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gw1.1.97.1\"</td>\n",
       "      <td>proteinId 46269</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>-1</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2098</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2883761.0</td>\n",
       "      <td>2884343.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gw1.1.99.1\"</td>\n",
       "      <td>proteinId 46443</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>2883734</td>\n",
       "      <td>False</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2099 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        scaffold  jgi type      start        end direction1 direction2 phase  \\\n",
       "0     scaffold_1  JGI  CDS  1273784.0  1273873.0          .          +     0   \n",
       "1     scaffold_1  JGI  CDS  1238222.0  1238786.0          .          +     0   \n",
       "2     scaffold_1  JGI  CDS  1999638.0  2000029.0          .          +     0   \n",
       "3     scaffold_1  JGI  CDS   857135.0   857702.0          .          +     0   \n",
       "4     scaffold_1  JGI  CDS  1189068.0  1189311.0          .          +     0   \n",
       "...          ...  ...  ...        ...        ...        ...        ...   ...   \n",
       "2094  scaffold_1  JGI  CDS  1394773.0  1395190.0          .          +     0   \n",
       "2095  scaffold_1  JGI  CDS  2885127.0  2885541.0          .          +     0   \n",
       "2096  scaffold_1  JGI  CDS  1898490.0  1898835.0          .          +     0   \n",
       "2097  scaffold_1  JGI  CDS   718868.0   719048.0          .          +     0   \n",
       "2098  scaffold_1  JGI  CDS  2883761.0  2884343.0          .          +     0   \n",
       "\n",
       "                      name         proteinId         n_exon  true_M_pos  \\\n",
       "0     name \"e_gw1.1.100.1\"   proteinId 60925   exonNumber 1          -1   \n",
       "1     name \"e_gw1.1.102.1\"   proteinId 61047   exonNumber 1          -1   \n",
       "2     name \"e_gw1.1.103.1\"   proteinId 61109   exonNumber 1     1999638   \n",
       "3     name \"e_gw1.1.105.1\"   proteinId 60920   exonNumber 1      857135   \n",
       "4     name \"e_gw1.1.106.1\"   proteinId 60980   exonNumber 1          -1   \n",
       "...                    ...               ...            ...         ...   \n",
       "2094     name \"gw1.1.94.1\"   proteinId 46209   exonNumber 1     1394773   \n",
       "2095     name \"gw1.1.95.1\"   proteinId 46224   exonNumber 1          -1   \n",
       "2096     name \"gw1.1.96.1\"   proteinId 46246   exonNumber 1     1898307   \n",
       "2097     name \"gw1.1.97.1\"   proteinId 46269   exonNumber 1          -1   \n",
       "2098     name \"gw1.1.99.1\"   proteinId 46443   exonNumber 1     2883734   \n",
       "\n",
       "       no_M  M_pos_diff  \n",
       "0      True           0  \n",
       "1      True           0  \n",
       "2     False           0  \n",
       "3     False           0  \n",
       "4      True           0  \n",
       "...     ...         ...  \n",
       "2094  False           0  \n",
       "2095   True           0  \n",
       "2096  False         183  \n",
       "2097   True           0  \n",
       "2098  False          27  \n",
       "\n",
       "[2099 rows x 14 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result[result['direction2'] == '+']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MDKTLKIRGNPKVFFDMEIGGEPAGRVVMQLRADVVPKTAENFRQLCLKEEGEGYKGSSFHRVIPGFMCQGGDFTNHNGTGGKSIYGEKFEDENFQLKHTGKGVLSMANAGPGTNGSQFFLCTAKTAWLD\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(Seq(emihu_seq_dict['scaffold_1'][\n",
    "    int(1999639.0-1):int(2000029.0)\n",
    "]).translate())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>scaffold</th>\n",
       "      <th>jgi</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>direction1</th>\n",
       "      <th>direction2</th>\n",
       "      <th>phase</th>\n",
       "      <th>name</th>\n",
       "      <th>proteinId</th>\n",
       "      <th>n_exon</th>\n",
       "      <th>true_M_pos</th>\n",
       "      <th>no_M</th>\n",
       "      <th>M_pos_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2099</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1234941.0</td>\n",
       "      <td>1235519.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>name \"e_gw1.1.1.1\"</td>\n",
       "      <td>proteinId 61078</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>1235144</td>\n",
       "      <td>False</td>\n",
       "      <td>375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2100</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>548869.0</td>\n",
       "      <td>549137.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>name \"e_gw1.1.104.1\"</td>\n",
       "      <td>proteinId 61169</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>548957</td>\n",
       "      <td>False</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2101</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>548869.0</td>\n",
       "      <td>549137.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>name \"e_gw1.1.107.1\"</td>\n",
       "      <td>proteinId 61038</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>548957</td>\n",
       "      <td>False</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2102</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>548869.0</td>\n",
       "      <td>549137.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>name \"e_gw1.1.108.1\"</td>\n",
       "      <td>proteinId 61099</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>548957</td>\n",
       "      <td>False</td>\n",
       "      <td>180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2103</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2686721.0</td>\n",
       "      <td>2687227.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>name \"e_gw1.1.110.1\"</td>\n",
       "      <td>proteinId 61045</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>2687227</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4190</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1456659.0</td>\n",
       "      <td>1456868.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>name \"gw1.1.87.1\"</td>\n",
       "      <td>proteinId 45700</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>1456784</td>\n",
       "      <td>False</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4191</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2843908.0</td>\n",
       "      <td>2844120.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>name \"gw1.1.89.1\"</td>\n",
       "      <td>proteinId 45750</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>2844069</td>\n",
       "      <td>False</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4192</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2705691.0</td>\n",
       "      <td>2705999.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>name \"gw1.1.90.1\"</td>\n",
       "      <td>proteinId 45766</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>2705984</td>\n",
       "      <td>False</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4193</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1629617.0</td>\n",
       "      <td>1629799.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>name \"gw1.1.92.1\"</td>\n",
       "      <td>proteinId 46190</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>1629625</td>\n",
       "      <td>False</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4194</th>\n",
       "      <td>scaffold_1</td>\n",
       "      <td>JGI</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1254079.0</td>\n",
       "      <td>1254451.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>name \"gw1.1.98.1\"</td>\n",
       "      <td>proteinId 46409</td>\n",
       "      <td>exonNumber 1</td>\n",
       "      <td>1254193</td>\n",
       "      <td>False</td>\n",
       "      <td>258</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2096 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        scaffold  jgi type      start        end direction1 direction2 phase  \\\n",
       "2099  scaffold_1  JGI  CDS  1234941.0  1235519.0          .          -     1   \n",
       "2100  scaffold_1  JGI  CDS   548869.0   549137.0          .          -     2   \n",
       "2101  scaffold_1  JGI  CDS   548869.0   549137.0          .          -     2   \n",
       "2102  scaffold_1  JGI  CDS   548869.0   549137.0          .          -     2   \n",
       "2103  scaffold_1  JGI  CDS  2686721.0  2687227.0          .          -     1   \n",
       "...          ...  ...  ...        ...        ...        ...        ...   ...   \n",
       "4190  scaffold_1  JGI  CDS  1456659.0  1456868.0          .          -     1   \n",
       "4191  scaffold_1  JGI  CDS  2843908.0  2844120.0          .          -     1   \n",
       "4192  scaffold_1  JGI  CDS  2705691.0  2705999.0          .          -     1   \n",
       "4193  scaffold_1  JGI  CDS  1629617.0  1629799.0          .          -     1   \n",
       "4194  scaffold_1  JGI  CDS  1254079.0  1254451.0          .          -     0   \n",
       "\n",
       "                      name         proteinId         n_exon  true_M_pos  \\\n",
       "2099    name \"e_gw1.1.1.1\"   proteinId 61078   exonNumber 1     1235144   \n",
       "2100  name \"e_gw1.1.104.1\"   proteinId 61169   exonNumber 1      548957   \n",
       "2101  name \"e_gw1.1.107.1\"   proteinId 61038   exonNumber 1      548957   \n",
       "2102  name \"e_gw1.1.108.1\"   proteinId 61099   exonNumber 1      548957   \n",
       "2103  name \"e_gw1.1.110.1\"   proteinId 61045   exonNumber 1     2687227   \n",
       "...                    ...               ...            ...         ...   \n",
       "4190     name \"gw1.1.87.1\"   proteinId 45700   exonNumber 1     1456784   \n",
       "4191     name \"gw1.1.89.1\"   proteinId 45750   exonNumber 1     2844069   \n",
       "4192     name \"gw1.1.90.1\"   proteinId 45766   exonNumber 1     2705984   \n",
       "4193     name \"gw1.1.92.1\"   proteinId 46190   exonNumber 1     1629625   \n",
       "4194     name \"gw1.1.98.1\"   proteinId 46409   exonNumber 1     1254193   \n",
       "\n",
       "       no_M  M_pos_diff  \n",
       "2099  False         375  \n",
       "2100  False         180  \n",
       "2101  False         180  \n",
       "2102  False         180  \n",
       "2103  False           0  \n",
       "...     ...         ...  \n",
       "4190  False          84  \n",
       "4191  False          51  \n",
       "4192  False          15  \n",
       "4193  False         174  \n",
       "4194  False         258  \n",
       "\n",
       "[2096 rows x 14 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result[result['direction2'] == '-']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSLGGEGRGQFDSAIDAAYDAGVLTVVAAG\n"
     ]
    }
   ],
   "source": [
    "print(Seq(emihu_seq_dict['scaffold_1'][\n",
    "    int(548869.0-3):int(548957.0)\n",
    "]).reverse_complement().translate())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSLGGEGRGQFDSAIDAAYDAGVLTVVAAG\n",
      "548957\n",
      "MSLGGEGRGQFDSAIDAAYDAGVLTVVAA\n",
      "548956\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tranquocan.truong/Documents/AnTTQ/U_Sorbonne/M1S2/Projet/projectS1/lib/python3.9/site-packages/Bio/Seq.py:2879: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "cds_start = 548869.0\n",
    "cds_end = 549137.0\n",
    "m_pos = reVerifyM(emihu_seq_dict['scaffold_1'],\n",
    "      cds_start,\n",
    "      cds_end,\n",
    "      '-'\n",
    ")\n",
    "print(m_pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSLGGEGRGQFDSAIDAAYDAGVLTVVAAG\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tranquocan.truong/Documents/AnTTQ/U_Sorbonne/M1S2/Projet/projectS1/lib/python3.9/site-packages/Bio/Seq.py:2879: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "print(Seq(emihu_seq_dict['scaffold_1'][\n",
    "    int(548869.0-3):int(548957)\n",
    "]).reverse_complement().translate())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GLLSNYEYLCHLNDAAGRSCADLAQYPVMPWVLQDYTSHTLDLADPAVYRDLSKPVGALDASRLALFRE\n"
     ]
    }
   ],
   "source": [
    "# cds_pos = int(result.loc[result['name'] == 'name \"gw1.1.87.1\"', 'start'].iloc[0])\n",
    "# m_pos = reVerifyM(emihu_seq_dict[result.loc[result['name'] == 'name \"gw1.1.87.1\"', 'scaffold'].iloc[0]],\n",
    "#       cds_pos=cds_pos,\n",
    "#       direction=result.loc[result['name'] == 'name \"gw1.1.87.1\"', 'direction2'].iloc[0]\n",
    "# )\n",
    "# print(cds_pos)\n",
    "# print(m_pos)\n",
    "\n",
    "print(Seq(emihu_seq_dict['scaffold_1'][\n",
    "    int(1456659.0):int(1456868.0)\n",
    "]).reverse_complement().translate())\n",
    "\n",
    "# Seq(emihu_seq_dict[result.loc[result['name'] == 'name \"e_gw1.1.1.1\"', 'scaffold'].iloc[0]][\n",
    "#     m_pos:int(result.loc[result['name'] == 'name \"e_gw1.1.1.1\"', 'end'].iloc[0])\n",
    "# ]).translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# testing e_gw1.1.100.1: why does it not start with M but is_diff = 0\n",
    "M_pos = 0\n",
    "stop_flag = False\n",
    "tmp_seq =  str(Seq(emihu_seq_dict['scaffold_1'][:1273784]).translate())[::-1]\n",
    "print(str(Seq(emihu_seq_dict['scaffold_1'][1273784:1273873]).translate()))\n",
    "for c in range(len(tmp_seq)):\n",
    "    print(tmp_seq[c])\n",
    "    if tmp_seq[c] == \"M\" and not stop_flag:\n",
    "        M_pos = c\n",
    "    if tmp_seq[c] == '*':\n",
    "        break\n",
    "\n",
    "M_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_pos = int(result[result['name'] == 'name \"e_gw1.1.100.1\"']['start'])\n",
    "m_pos = reVerifyM(emihu_seq_dict[result[result['name'] == 'name \"e_gw1.1.100.1\"']['scaffold'].iloc[0]],\n",
    "      cds_pos=cds_pos\n",
    ")\n",
    "Seq(emihu_seq_dict[result[result['name'] == 'name \"e_gw1.1.100.1\"']['scaffold'].iloc[0]][\n",
    "    m_pos:int(result[result['name'] == 'name \"e_gw1.1.100.1\"']['end'])\n",
    "]).translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result[result['is_diff'] != 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "43566 / 228000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# packaging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "\n",
    "\n",
    "def load_gff(file_path:str ='./Emihu1_all_genes.gff', \n",
    "             columns:list=['scaffold', 'jgi', 'type', 'start', 'end', 'direction1', 'direction2', 'phase', 'infos']\n",
    "             ) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Read table from .GFF file and load in to Pandas.DataFrame table\n",
    "    \"\"\"\n",
    "    # check path\n",
    "    if not os.path.exists(file_path):\n",
    "        print(\"Non existed input path, Please re-check!\")\n",
    "        return None\n",
    "    \n",
    "    # Load the GFF file\n",
    "    df_gff = pd.read_csv(file_path, sep='\\t', comment='#', names=columns)\n",
    "\n",
    "    # Split the 'attributes' column by ';' into multiple columns\n",
    "    df_gff_split = df_gff['infos'].str.split(';', expand=True)\n",
    "    # Rename the new columns (optional)\n",
    "    df_gff_split.columns = ['name', 'proteinId', 'n_exon']\n",
    "    df_gff.columns = df_gff.columns.str.strip()\n",
    "\n",
    "    # Add the split columns back to the original DataFrame\n",
    "    df_gff = pd.concat([df_gff, df_gff_split], axis=1)\n",
    "    df_gff.drop(['infos'], axis=1, inplace=True)\n",
    "    df_gff = df_gff[~df_gff['proteinId'].isna()]\n",
    "    # convert CDS position to integer and subtract by 1 because string start from 0 in python\n",
    "    df_gff['start'] = df_gff['start'].astype(int) - 1\n",
    "    \n",
    "    return df_gff\n",
    "\n",
    "\n",
    "def get_Coordinates(\n",
    "        df_gff:pd.DataFrame, \n",
    "        groupby_attr:str ='name'\n",
    "        ) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Filtering the first CDS coordinate infomation for each attribute ('Name' by default) group\n",
    "    from the GFF dataframe \n",
    "    \"\"\"\n",
    "\n",
    "    # if not df_gff.isinstance(pd.DataFrame):\n",
    "    #     print(\"Non existed input path, Please re-check!\")\n",
    "    #     return None\n",
    "    \n",
    "    # Group by 'group' and find the first row with a non-null value\n",
    "    result = df_gff.groupby(groupby_attr, group_keys=False).apply(\n",
    "        lambda group: group.loc[group[group['type'] == 'CDS'].index[0]] if any(group['type'] == 'CDS') else None,\n",
    "        include_groups=False\n",
    "    )\n",
    "\n",
    "    return result\n",
    "\n",
    "def load_Fasta(fasta_path:str = './Emihu1_masked_scaffolds.fasta'):\n",
    "    \n",
    "    if not os.path.exists(fasta_path):\n",
    "        print(\"Non existed input path, Please re-check!\")\n",
    "        return None\n",
    "    \n",
    "    ret_dict = {}\n",
    "    key = ''\n",
    "    with open(fasta_path, 'r') as f:\n",
    "        for line in f.readlines():\n",
    "            if \">\" in line:\n",
    "                # translate acide amine sequence to protein before advance\n",
    "                if key != '':\n",
    "                    ret_dict[key] = ret_dict[key]\n",
    "                # save scaffold number as dict key\n",
    "                key = line[1:].strip()  # avoid '>'\n",
    "                # create a value for storing the sequence\n",
    "                ret_dict[key] = ''\n",
    "            else:\n",
    "                ret_dict[key] += line.strip()\n",
    "    f.close()\n",
    "\n",
    "    return ret_dict\n",
    "\n",
    "\n",
    "def reVerifyM(\n",
    "        fasta_seq: str, \n",
    "        cds_pos: float\n",
    "        ) -> int:\n",
    "    \"\"\"\n",
    "    Take an input of fasta acide amine sequence, CDS position from JGI \n",
    "    and re-find the M position in the given sequence\n",
    "    \"\"\"\n",
    "    cds_pos = int(cds_pos)\n",
    "    if cds_pos < 0 or cds_pos >= len(fasta_seq):\n",
    "        return cds_pos\n",
    "\n",
    "    # include the first translated codon\n",
    "    sub_seq = fasta_seq[:cds_pos+3]\n",
    "    # print(len(sub_seq))\n",
    "    # identify the translation framework: +0,+1 or +2 depend on the length of sequence\n",
    "    framework = len(sub_seq)%3\n",
    "    sub_seq = str(Seq(sub_seq[framework:]).translate())\n",
    "    # print(len(sub_seq))\n",
    "    \n",
    "    # find the closest codon stop\n",
    "    star_indices = [m.start() for m in re.finditer(r'\\*', sub_seq)]\n",
    "    if not star_indices:\n",
    "        return cds_pos\n",
    "    closest_stop = min(star_indices, key=lambda i: abs(i - cds_pos))\n",
    "    # print(closest_stop)\n",
    "    stop_pos = closest_stop*3 + framework\n",
    "    # print(stop_pos)\n",
    "\n",
    "    # find the closest aa M to the closest codon stop *\n",
    "    sub_seq = sub_seq[closest_stop:]\n",
    "    star_indices = [m.start() for m in re.finditer(r'M', sub_seq)]\n",
    "    if not star_indices:\n",
    "        return -1 # mean there is no M from the position cds+1 to the closest *\n",
    "    closest_M = min(star_indices, key=lambda i: abs(i - closest_stop))\n",
    "    M_pos = closest_M*3 + stop_pos\n",
    "    # print(M_pos)\n",
    "    # print(Seq(fasta_seq[M_pos:cds_pos+3]).translate())\n",
    "    # sub_seq = sub_seq[closest_M:]\n",
    "\n",
    "    # print(sub_seq)\n",
    "    # print(closest_stop)\n",
    "\n",
    "    return int(M_pos)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!! Why not translate DNA to protein before saving into dict?\n",
    "=> To respect the framework of translation (+1, +2, +3). If we translate before saving, we can translate the wrong framework compare to the true translated one then we'll translate after getting the sub-sequence before the CDS Position. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gff_emihu1 = load_gff(emihu_gff)\n",
    "fasta_dict_emihu1 = load_Fasta(emihu_seq_path)\n",
    "df_cds_emihu1 = get_Coordinates(df_gff_emihu1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gff_emihu1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cds_emihu1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "type(df_cds_emihu1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cds_emihu1[df_cds_emihu1.index == 'name \"gw1.13.147.1\"']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = fasta_dict_emihu1['scaffold_13'][\n",
    "    int(df_cds_emihu1[df_cds_emihu1.index == 'name \"gw1.13.147.1\"']['start']):int(df_cds_emihu1[df_cds_emihu1.index == 'name \"gw1.13.147.1\"']['end']+1)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(tmp), tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "! ALL GOOD"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "projectS1",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
